### PR TITLE
Add typescript infrastructure and convert codebase

### LINKED ISSUE #4: Add typescript support
Add TS support and convert source code to TS.

### PR DESCRIPTION
**Pull request by zakangelle**
This PR adds all the necessary tooling infrastructure to support TypeScript, including:

- Install and configure `typescript` library
- Install `@types` libraries
- Add `tsc` command
- Type all `src` and `example` modules
- Update babel config to handle TS files
- Update webpack `example` build to handle TS files
- Update eslint config to handle TS files
- Add jest config to handle TS files

An example consumer implementation:

```ts
interface Data {
  data: {
    name: string;
    color: string;
  };
}

export default () => {
  const { data, loading, error } = useFetch<Data>({
    url: 'https://reqres.in/api/things/1',
  });

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>{error.message}</div>;
  }

  return (
    <>
      {data && (
        <div>
          {data.data.name}
          {': '}
          {data.data.color}
        </div>
      )}
    </>
  );
};
```

Closes #4

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 bin
 package-lock.json
-node_modules
\ No newline at end of file
+node_modules
+webpack.config.js
```
Comment (jpangelle) [inline .eslintignore]: Why can't the webpack file be linted?

File: README.md [hunk 2]
```diff
@@ -16,7 +16,9 @@ $ npm install use-fetch-hooks
 import { useFetch } from 'use-fetch-hooks';
 
 export default () => {
-  const { data, error, loading } = useFetch('https://example/api');
+  const { data, error, loading } = useFetch({
```
Comment (jpangelle) [inline README.md]: Do we not plan to accept just the default anymore? 

```js
useFetch('https://example/api');
```

File: README.md [hunk 3]
```diff
@@ -61,6 +63,45 @@ export default () => {
...
+
+interface Data {
+  data: {
+    name: string;
+    color: string;
+  };
+}
+
+export default () => {
+  const { data, loading, error } = useFetch<Data>({
```
Comment (jpangelle) [inline README.md]: Wanna keep this consistent through our codebase 😬 

alpha
```ts
const { data, error, loading } = useFetch<Data>({
    url: 'https://example/api',
  });
```

File: example/src/useFetch/Basic.tsx [hunk 4]
```diff
@@ -1,8 +1,17 @@
...
+interface Data {
+  data: {
+    name: string;
+    color: string;
+  };
+}
+
 export default () => {
-  const { data, error, loading } = useFetch('https://reqres.in/api/things/1');
+  const { data, loading, error } = useFetch<Data>({
```
Comment (jpangelle) [inline example/src/useFetch/Basic.tsx]: alpha here as well:
```ts
const { data, error, loading } = useFetch<Data>({
    url: 'https://reqres.in/api/things/1',
  });
```

File: example/src/useLazyFetch/Basic.tsx [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
+    name: string;
+    color: string;
+  };
+}
+
 export default () => {
-  const [getData, { data, error, loading }] = useLazyFetch(
-    'https://reqres.in/api/things/1'
-  );
+  const [getData, { data, loading, error }] = useLazyFetch<Data>({
```
Comment (jpangelle) [inline example/src/useLazyFetch/Basic.tsx]: alph

File: README.md (hunk 2 above)
Comment (zakangelle) [reply README.md]: Yeah, I was actually having trouble typing this properly. It's expressed in `axios` types [here](https://github.com/axios/axios/blob/master/index.d.ts#L125), but I can't figure out how to express an overloaded function signature that matches that.

Comment (zakangelle) [reply README.md]: We can revisit this soon.

File: README.md (hunk 3 above)
Comment (zakangelle) [reply README.md]: If that's the case it should be a lint/prettier config.

Comment (jpangelle) [reply README.md]: Hm, would have to look into to see if that's possible

File: .eslintignore (hunk 1 above)
Comment (zakangelle) [reply .eslintignore]: Updated so it's linting it again - had to add a couple lint overrides but should be good 👍

File: README.md (hunk 3 above)
Comment (zakangelle) [reply README.md]: Updated those examples to use alphabetical destructuring.

Could use something like [this](https://www.npmjs.com/package/eslint-plugin-sort-destructure-keys) if you want to enforce this via linting.

IMO it's kinda counterintuitive - feel like destructured properties should be ordered by importance vs. alphabetically.

**Pull request merged**